gson/src/main/java/com/google/gson/TypeAdapter.java

```diff
@@ -289,24 +289,30 @@ public final T fromJsonTree(JsonElement jsonTree) {
    * Note that we didn't need to check for nulls in our type adapter after we used nullSafe.
    */
   public final TypeAdapter<T> nullSafe() {
-    return new TypeAdapter<T>() {
-      @Override
-      public void write(JsonWriter out, T value) throws IOException {
-        if (value == null) {
-          out.nullValue();
-        } else {
-          TypeAdapter.this.write(out, value);
-        }
+    if (this.getClass() != NullSafeTypeAdapter.class) {
```

Comment:
Not using `instanceof` may look weird, but doesn't `instanceof` cause a compile time error, or am I missing something?